### PR TITLE
517 add factory for customized functionalities

### PR DESCRIPTION
**Pull request by mathpluscode**
# Description

Use registry for backbone and refactor backbone related configs.

## Type of change

What types of changes does your code introduce to DeepReg? _Put an `x` in the boxes that
apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [x] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [x] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [x] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by mathpluscode** [comment]
I noticed that function `build_model` was not well covered by tests. But it's fine to let it go now, as later we will use the registry for the models as well. Code will be covered in the next PR.

**Comment by mathpluscode** [comment]
This PR should be ready now. @YipengHu @NMontanaBrown @RemiDelaunay would you have time to review this :D thx!

**Comment by mathpluscode** [comment]
BTW, the demo configs have been changed but I've tested them locally and in CI. So they are safe ;)

**Comment by mathpluscode** [comment]
@acasamitjana Please have a look as I'm changing the config ;) Also feel free to check the issue #517.

**Comment by NMontanaBrown** [comment]
I can review this but it will take me a while as there are 33 files. Over the weekend maybe

File: deepreg/model/network/build.py
Comment (NMontanaBrown) [inline deepreg/model/network/build.py:22]: Can you make the description consistent between files? "config for the model and loss" as it was referred to before.

**Review by NMontanaBrown — COMMENTED** [review]
Generally looks good but since we add the registry method, could you add a docs entry for registry explaining functionality (as this seems rather key and useful new feature), and raise a ticket for demo for using registry? The unit test that you wrote is useful to illustrate but again users would benefit from illustration of new feature.

**Comment by mathpluscode** [comment]
> Generally looks good but since we add the registry method, could you add a docs entry for registry explaining functionality (as this seems rather key and useful new feature), and raise a ticket for demo for using registry? The unit test that you wrote is useful to illustrate but again users would benefit from illustration of new feature.

I'm considering a doc in the end as I'm planning to pass a 'registry' to the train/predict method, which by default will be ´Registry()´, but user can register extra things.

Regarding demos, I think they are ok as it's using existing methods.

Will add one more folder examples, to provide self contained scripts showing how to use registry and docs will be there in RTD.

But I will do the doc in the end as it was meant to be one huge PR but I guess splitting it will be easier for code review.

So I will add a doc request issue. For now user actually do not feel the difference except the config changes:)

**Comment by mathpluscode** [comment]
@RemiDelaunay @acasamitjana Would you have time to review this ;) I'm waiting for approval so that I can carry on more following changes.

**Comment by NMontanaBrown** [comment]
@mathpluscode where is the docs request issue for his feature? or is it going to be added to this issue?

**Comment by mathpluscode** [comment]
> @mathpluscode where is the docs request issue for this feature? or is it going to be added to this issue?

It's still in issue #517 but this PR is not closing that issue.

**Review by NMontanaBrown — APPROVED** [review]

**Review by acasamitjana — COMMENTED** [review]
Hey, very nice, thank you.
I was just wondering whether we can arrange also the loss part in the config file (e.g., remove the level dissimilarity-regularization). Also allowing multiple loss functions in relation to #526 

I'm sure there is a better way, but as an example, if we want a SSD and NCC as image dissimilarity losses:
image:
   name: ['SSD','NCC']
   SSD: 
      weight: 2.0
   NCC: 
       weight: 1.0
      *other_NCC_parameters:

**Comment by mathpluscode** [comment]
> Hey, very nice, thank you.
> 
> I was just wondering whether we can arrange also the loss part in the config file (e.g., remove the level dissimilarity-regularization). Also allowing multiple loss functions in relation to #526 
> 
> 
> 
> I'm sure there is a better way, but as an example, if we want a SSD and NCC as image dissimilarity losses:
> 
> image:
> 
> - name: ['SSD','NCC']
> 
> - SSD: 
> 
> --- weight: 2.0
> 
> - NCC: 
> 
> --- weight: 1.0
> 
> --- *other_NCC_parameters:
> 
> 

Yes! We should definitely do like this. It can be done in a separate MR. The idea would be passing a 'registry' at the very beginning like before the training/evaluation.

😛if you are interested in this, we can share the workload.

**Comment by acasamitjana** [comment]
> 😛if you are interested in this, we can share the workload.
> 
> also what you are saying is linked to #526

Okey, let's do that there.

**Review by acasamitjana — APPROVED** [review]

**Pull request merged**
